src/qibo/models/encodings.py

```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
+                "tier": weight,
+                "children": [],
+            }
+            for it in range(nqubits // weight)
+        ]
+        nmodk = nqubits % weight
+        if nmodk != 0:
+            disjoint_sets.append(
+                {
+                    "qubits": list(range(nqubits - nmodk, nqubits)),
+                    "tier": nmodk,
+                    "children": [],
+                }
+            )
+        disjoint_sets = list(
+            reversed(disjoint_sets)
+        )  # we reverse to have in ascending order of tier
+
+        trees = disjoint_sets.copy()
+        # combine lowest tier trees into one tree
+        while len(trees) > 1:
+            first_smallest = trees.pop(0)
+            second_smallest = trees.pop(0)
+            second_smallest["tier"] += first_smallest["tier"]
+            second_smallest["children"].append(first_smallest)
+            new = second_smallest
+            # put new combined tree in list mantaining ordering
+            trees.insert(sum(x["tier"] < new["tier"] for x in trees), new)
+
+        root = trees[0]
+        # We initialize |0>^|1>^k  bitstring at root qubits
+        circuit.add(gates.X(q) for q in root["qubits"][-weight:])
+
+        # undo the union-by-tier algorithm:
+        # split each root's (x) highest tier child y, and add WBD acting on both
+        while len(trees) < len(disjoint_sets):
+            cut_nodes = []
+            for node in trees:
+                if len(node["children"]) > 0:
+                    # find highest tier child
+                    y = max(node["children"], key=lambda x: x["tier"])
+                    # add WBD acting on both sets of qubits
+                    _add_wbd_gate(
+                        circuit,
+                        node["qubits"],
+                        y["qubits"],
+                        node["tier"],
+                        y["tier"],
+                        weight,
+                    )
+                    # cut tree splitting x and y
+                    node["tier"] -= y["tier"]
+                    node["children"].remove(y)
+                    cut_nodes.append(y)
+            trees += cut_nodes
+
+        for node in disjoint_sets:
+            _add_dicke_unitary_gate(
+                circuit, node["qubits"], min(weight, len(node["qubits"]))
+            )
 
     return circuit
 
 
-def _add_scs_gate(circuit: Circuit, n: int, k: int):
+def _add_scs_gate(circuit: Circuit, qubits: List[int], k: int):
     """In-place addition of a Split & Cyclic Shift (SCS) gate to ``circuit``.
-
-    Implements the SCS_{n,k} unitary from Definition 3 of the paper.
-    Acts on the last k+1 qubits of the circuit.
+    Implements the SCS_{n,k} unitary from Definition 3 of the paper [1].
+    Acts on the last k+1 qubits of the n passed qubits.
     """
     if k == 0:
         return  # SCS_{n,0} is identity
 
-    last_qubit = n - 1
-    # first_qubit = last_qubit - k
+    n = len(qubits)
```

Comment:
```suggestion
    nqubits = len(qubits)
```

Avoiding non-descriptive variable names